src/Nri/Ui/Highlighter/V5.elm

```diff
@@ -1123,6 +1112,155 @@ staticMarkdownWithTags =
         )
 
 
+findOverlapsSupport : Model marker -> OverlapsSupport marker
+findOverlapsSupport model =
+    let
+        hoveredMarkers =
+            model.highlightables
+                |> List.Extra.find (\h -> Just h.index == model.mouseOverIndex)
+                |> Maybe.map (.marked >> List.map .kind)
+                |> Maybe.withDefault []
+    in
+    OverlapsSupported
+        { hoveredMarkerWithShortestHighlight =
+            case hoveredMarkers of
+                [] ->
+                    Nothing
+
+                marker :: [] ->
+                    Just marker
+
+                first :: second :: rest ->
+                    Just
+                        (markerWithShortestHighlight
+                            model.sorter
+                            model.highlightables
+                            ( first, second, rest )
+                        )
+        }
+
+
+{-| A type that contains information needed to render individual `Highlightable`s one at a time
+-}
+type FoldState marker
+    = FoldState
+        { model : Model marker
+        , overlapsSupport : OverlapsSupport marker
+        , state : List ( Highlightable marker, Maybe (Html Never), List Css.Style )
+        }
+
+
+{-| Computes all the mark styles necessary to perform a fold over the `Highlightable` elements
+-}
+initFoldState : Model marker -> FoldState marker
+initFoldState model =
+    let
+        overlapsSupport =
+            findOverlapsSupport model
+
+        config =
+            { hintingIndices = model.hintingIndices
+            , mouseOverIndex = model.mouseOverIndex
+            , mouseDownIndex = model.mouseDownIndex
+            , maybeTool = Just model.marker
+            , overlaps = overlapsSupport
+            , highlightables = model.highlightables
+            }
+
+        highlightableGroups =
+            buildGroups config model.highlightables
+
+        toMark : Highlightable marker -> Tool.MarkerModel marker -> Mark.Mark
+        toMark highlightable marker =
+            { name = marker.name
+            , startStyles = marker.startGroupClass
+            , styles =
+                markedHighlightableStyles config
+                    (isHovered_ config highlightableGroups)
+                    highlightable
+            , endStyles = marker.endGroupClass
+            }
+
+        markStyles =
+            model.highlightables
+                |> List.map (\highlightable -> ( highlightable, List.map (toMark highlightable) highlightable.marked ))
+                |> Mark.overlappingStyles
+    in
+    FoldState
+        { model = model
+        , overlapsSupport = overlapsSupport
+        , state = markStyles
```

Comment:
💭 reading this line it seems pretty evident to me that `markStyles` and `overlappingStyles` really downplay the importance of we are doing here! like... when i read it my mind went "oh just computes a bunch of styles" when in reality we're building an intermediate representation that is basically the model we'll be rendering.

sharing a thought without a suggestion because TBH i don't know how I would improve it. feel free to ignore this 😛